src/Microsoft.Identity.Web.OidcFIC/OidcIdpSignedAssertionProvider.cs

```diff
@@ -17,21 +19,58 @@ internal class OidcIdpSignedAssertionProvider : ClientAssertionProviderBase
         private readonly ITokenAcquirerFactory _tokenAcquirerFactory;
         private readonly MicrosoftIdentityApplicationOptions _options;
         private readonly string? _tokenExchangeUrl;
+        private readonly ILogger? _logger;
 
-        public OidcIdpSignedAssertionProvider(ITokenAcquirerFactory tokenAcquirerFactory, MicrosoftIdentityApplicationOptions options, string? tokenExchangeUrl)
+        public bool RequiresSignedAssertionFmiPath { get; internal set; }
+
+        public OidcIdpSignedAssertionProvider(ITokenAcquirerFactory tokenAcquirerFactory, MicrosoftIdentityApplicationOptions options, string? tokenExchangeUrl, ILogger?   logger)
         {
             _tokenAcquirerFactory = tokenAcquirerFactory;
             _options = options;
             _tokenExchangeUrl = tokenExchangeUrl;
+            _logger = logger;
         }
 
         protected override async Task<ClientAssertion> GetClientAssertionAsync(AssertionRequestOptions? assertionRequestOptions)
         {
             _tokenAcquirer ??= _tokenAcquirerFactory.GetTokenAcquirer(_options);
 
             string tokenExchangeUrl = _tokenExchangeUrl ?? "api://AzureADTokenExchange";
+            AcquireTokenOptions? acquireTokenOptions = null;
+
+            // During the construction of the CCA, IdWeb tried to understand which Credential description to use and to skip, and thefore
+            // attempts to load the credentials with assertionRequestOptions = null (whereas it's not null when MSAL calls GetSignedAssertionAsync).
+            // If assertionRequestOptions = null and RequiresSignedAssertionFmiPath is true
+            // we postpone getting the signed assertion until the first call, when ClientAssertionFmiPath will be provided.
+            if (RequiresSignedAssertionFmiPath && assertionRequestOptions == null)
+            {
+                if (_logger != null)
+                {
+                    _logger.LogDebug("OidcIdpSignedAssertionProvider: RequiresSignedAssertionFmiPath is true, but assertionRequestOptions is null. Postponing to first call");
+                }
 
-            AcquireTokenResult result = await _tokenAcquirer.GetTokenForAppAsync(tokenExchangeUrl + "/.default");
+                // By using Now, we are certain to be called immediately again
+                return new ClientAssertion(null!, DateTimeOffset.Now);
+            }
+
+
+            if (assertionRequestOptions != null && !string.IsNullOrEmpty(assertionRequestOptions.ClientAssertionFmiPath))
+            {
+                acquireTokenOptions = new AcquireTokenOptions()
+                {
+                    FmiPath = assertionRequestOptions.ClientAssertionFmiPath
+                };
+            }
+
+            if (_logger != null)
```

Comment:
same comment as above, and shouldn't this use the high performant logging?

Reply:
I'll create a github issue for that./